### PR TITLE
RATIS-997: Benchmarking protobuffers and flatbuffers for streaming in GRPC

### PR DESCRIPTION
**Pull request by anshkhannasbu**
## What changes were proposed in this pull request?

 Benchmarking protobuffers and flatbuffers for streaming in GRPC

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-997

## How was this patch tested?

Ran the client server program for 100,000 message transfers.

### CODE REVIEW

**Comment by szetszwo** [comment]
Thanks a lot for testing flatbuffer!  Some comments:
- Let’s move the program to a new module “ratis-experimental”.  By default, the module won’t be compiled.  It will be enabled by a profile.
- Please add some simple javadoc to describe the program.

File: ratis-examples/README.md [hunk 1]
```diff
@@ -138,6 +138,21 @@ for server: `java -cp target/*.jar org.apache.ratis.examples.counter.server.Coun
 replace {serverIndex} with 1, 2, or 3
 for client: `java -cp target/*.jar org.apache.ratis.examples.counter.client.CounterClient`
 
+## Example 4: Data Transfer
```
Comment (avijayanhwx) [inline ratis-examples/README.md:141]: This needs to be removed.

File: ratis-experiments/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+-->
+
+# Apache Ratis Experiments
+
+#### Building and Running The Examples
+
+The repository can be complied using `mvn clean package -DskipTests -DbuildExp` under the project root directory;
+see also [BUILDING.md](../BUILDING.md).
+
+This module contains the experiments carried out while building the project.
```
Comment (avijayanhwx) [inline ratis-experiments/README.md:22]: nit. This module has been created to house experimental projects.

File: ratis-experiments/pom.xml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            <artifactId>ratis-thirdparty-misc</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.ratis</groupId>
+            <artifactId>ratis-proto</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.5</version>
```
Comment (avijayanhwx) [inline ratis-experiments/pom.xml:38]: Can we check if we can use <commons.lang.version> defined in parent pom?

**Comment by szetszwo** [comment]
Thanks for the update.  Some comments on the changes.
- Revert the changes in ratis-examples/pom.xml, README.md
- In ratis-experiments/README.md,
-* it incorrectly refers to "examlpes".
-* Replace "project issue" with "flatbuffers issue #6023".
- Move FileTransfer.fbs and flatbuf-compile.sh to ratis-experiments.
- Move the protos from Examples.proto to a new file Experiments.proto.
- Use 2-space indentation in ratis-experiments/pom.xml

File: ratis-experiments/src/main/java/org/apache/ratis/experiments/flatbuffers/client/ClientFlat.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import org.apache.ratis.flatbufs.TransferMsg;
+import org.apache.ratis.flatbufs.TransferReply;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+
+
```
Comment (avijayanhwx) [inline ratis-experiments/src/main/java/org/apache/ratis/experiments/flatbuffers/client/ClientFlat.java:37]: nit. Java doc.

File: ratis-proto/src/main/proto/Examples.proto [hunk 5]
```diff
@@ -65,4 +65,4 @@ message WriteReplyProto {
 
 message DeleteReplyProto {
   bytes resolvedPath = 1;
-}
+}
```
Comment (szetszwo) [inline ratis-proto/src/main/proto/Examples.proto:68]: Please revert this.

File: ratis-proto/src/main/proto/Experiments.proto [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+message TransferReplyProto {
+    uint32 partId = 1;
+    string message = 2;
+}
+
+service FileTransferExampleService {
+    rpc sendData(stream TransferMsgProto) returns (stream TransferReplyProto);
+}
```
Comment (amaliujia) [inline ratis-proto/src/main/proto/Experiments.proto:36]: Need newline?

**Pull request merged**